platform-api/che-core-api-account/src/main/java/org/eclipse/che/api/account/server/AccountWorkspaceHooks.java

```diff
@@ -0,0 +1,143 @@
+        this.accountDao = accountDao;
+    }
+
+    /**
+     * Rejects workspace start if it is impossible to determine account
+     * which <i>is/should be</i> related with running workspace.
+     *
+     * <p>When workspace is not added to account yet, tries to detect an account
+     * and add workspace to it.
+     *
+     * <p>Reject/Allow scheme:
+     * <pre>
+     *  Is accountId specified ?
+     *      NO: Is workspace already added to any account?
+     *          YES: workspace start is <b>allowed</b>
+     *          NO: Does user have a single account?
+     *              YES: add workspace to account, start is <b>allowed</b>
+     *              NO: workspace start is <b>rejected</b>
+     *      YES: Is workspace already added to any account?
+     *          YES: Does this account id equal to specified accountId?
+     *              YES: workspace start is <b>allowed</b>
+     *              NO: workspace start is <b>rejected</b>
+     *          NO: Is the current user an account owner of the specified account?
+     *              YES: add workspace to account, start is <b>allowed</b>
+     *              NO: workspace start is <b>rejected</b>
+     * </pre>
+     */
+    @Override
+    public void beforeStart(@NotNull UsersWorkspace workspace, @Nullable String accountId) throws NotFoundException,
+                                                                                                  ForbiddenException,
+                                                                                                  ServerException {
+        User currentUser = EnvironmentContext.getCurrent().getUser();
+        if (accountId == null) {
+            // check if account is already specified for given workspace
+            if (!accountDao.isWorkspaceRegistered(workspace.getId())) {
+                // when workspace is not added to any of accounts
+                // try to detect whether current user is owner of a single account
+                List<Account> ownedAccounts = accountDao.getByOwner(currentUser.getId());
+                // there is no way to get empty list, user always has at least one account
+                if (ownedAccounts.size() != 1) {
+                    throw new ForbiddenException(format("Workspace start rejected. Impossible to determine account for workspace '%s', "
+                                                        + "user '%s' is owner of several accounts. Specify account identifier!",
+                                                        workspace.getId(),
+                                                        workspace.getOwner()));
+                }
+                // account detection is completed
+                // if user is owner of a single account add workspace to it
+                Account account = ownedAccounts.get(0);
+                account.getWorkspaces().add(workspace);
+                accountDao.update(account);
```

Comment:
As far as i remember we decided to use both implicit and explicit approaches.

_The point for implicit:_ in 90% situations user is account/owner of a single account, then it is possible to detect it and use, which simplifies the usage of the API.
_The point for explicit:_ when user explicitly specifies workspace it is obvious and has no side effects.

Actually i agree with you and i think that workspace registration should be done in explicit way.

Btw all cases are described [here](https://github.com/codenvy/che-core/blob/IDEX-2812/platform-api/che-core-api-account/src/main/java/org/eclipse/che/api/account/server/AccountWorkspaceHooks.java#L57-L70)